engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationMigrateProcessor.java

```diff
@@ -64,19 +75,68 @@ public void processRecord(final TypedRecord<ProcessInstanceMigrationRecord> comm
               value.getProcessInstanceKey(), targetProcessDefinitionKey));
     }
 
+    // avoid stackoverflow using a queue to iterate over the descendants instead of recursion
+    final var elementInstances = new ArrayDeque<>(List.of(processInstance));
+    while (!elementInstances.isEmpty()) {
+      final var elementInstance = elementInstances.poll();
+      final List<ElementInstance> children =
+          migrateElementInstance(elementInstance, processDefinition, mappingInstructions);
+      elementInstances.addAll(children);
+    }
+
+    stateWriter.appendFollowUpEvent(
+        processInstanceKey, ProcessInstanceMigrationIntent.MIGRATED, value);
+    responseWriter.writeEventOnCommand(
+        processInstanceKey, ProcessInstanceMigrationIntent.MIGRATED, value, command);
+  }
+
+  private List<ElementInstance> migrateElementInstance(
+      final ElementInstance elementInstance,
+      final DeployedProcess processDefinition,
+      final List<ProcessInstanceMigrationMappingInstructionValue> mappingInstructions) {
+
+    final String targetElementId =
+        determineTargetElementId(elementInstance, processDefinition, mappingInstructions);
+
     stateWriter.appendFollowUpEvent(
-        processInstanceKey,
+        elementInstance.getKey(),
         ProcessInstanceIntent.ELEMENT_MIGRATED,
-        processInstance
+        elementInstance
             .getValue()
             .setProcessDefinitionKey(processDefinition.getKey())
             .setBpmnProcessId(processDefinition.getBpmnProcessId())
             .setVersion(processDefinition.getVersion())
-            .setElementId(processDefinition.getBpmnProcessId()));
+            .setElementId(targetElementId));
 
-    stateWriter.appendFollowUpEvent(
-        processInstanceKey, ProcessInstanceMigrationIntent.MIGRATED, value);
-    responseWriter.writeEventOnCommand(
-        processInstanceKey, ProcessInstanceMigrationIntent.MIGRATED, value, command);
+    if (elementInstance.getJobKey() > 0) {
```

Comment:
🔧 Again, would be good if we create a separate method for handling job migration. That will mitigate the complexity of the `migrateElementInstance` method.